### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by artemd24**

### CODE REVIEW

File: homework4/task_1_read_file.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+**** https://docs.python.org/3/tutorial/errors.html#raising-exceptions
+"""
+
+
+def read_magic_number(path: str) -> bool:
+    with open(path, "r", encoding="UTF-8") as file:
+        try:
+            return 1 <= float(file.readline()) < 3
+        except Exception:
+            raise ValueError
```
Comment (apaniukov) [inline homework4/task_1_read_file.py]: The type of the error is good, but you should also give some explanation in the error message too. Also, `float("abc")` already raises `ValueError`, so you just duplicated this logic and deleted the error message: 
```python
>>> float("a")
Traceback (most recent call last):
  File "<input>", line 1, in <module>
ValueError: could not convert string to float: 'a'
```

File: homework4/tests/task1_test.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    assert read_magic_number("test.txt") == expected_result
+    os.remove("test.txt")
+
+
+@pytest.mark.parametrize(
+    "length",
+    [randint(-10000, 10000)],
+)
+def test_negative_read_magic_number(length: int):
```
Comment (apaniukov) [inline homework4/tests/task1_test.py]: You don't have to use `parametrize` if you have only one value to check in the test.

File: homework4/task_3_get_print_output.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+"""
+import sys
+
+
+def my_precious_logger(text: str):
+    if text.startswith("error"):
+        print(text, file=sys.stderr)
+    else:
+        # print(text)
+        sys.stdout.write(text)
```
Comment (apaniukov) [inline homework4/task_3_get_print_output.py:31]: Why do you use different output methods? Is there any difference between `print` and `sys.stdout.write`?

File: homework4/task_5_optional.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+>>> list(fizzbuzz(5))
+['1', '2', 'fizz', '4', 'buzz']
+
+* https://en.wikipedia.org/wiki/Fizz_buzz
+** https://www.youtube.com/watch?v=NSzsYWckGd4
+"""
+from typing import Generator
+
+
+def fizzbuzz(n: int) -> Generator[str, str, str]:
```
Comment (apaniukov) [inline homework4/task_5_optional.py]: Technicaly, it should be `-> Generator[str, None, None]`.

File: homework4/task_1_read_file.py (hunk 1 above)
Comment (artemd24) [reply homework4/task_1_read_file.py]: I was just following the task which says _In case of **any** error, a ValueError should be thrown._, ValueError is not the one of possibles error, I guess stdin could raise _EOFError_

File: homework4/tests/task1_test.py (hunk 2 above)
Comment (artemd24) [reply homework4/tests/task1_test.py]: Done

File: homework4/task_3_get_print_output.py (hunk 3 above)
Comment (artemd24) [reply homework4/task_3_get_print_output.py:31]: Done

File: homework4/task_5_optional.py (hunk 4 above)
Comment (artemd24) [reply homework4/task_5_optional.py]: Done

**Review by apaniukov — APPROVED** [review]

**Pull request merged**